core/shared/platform/windows/win_file.c

```diff
@@ -4,117 +4,766 @@
  */
 
 #include "platform_api_extension.h"
-#include "platform_internal.h"
+#include "libc_errno.h"
+#include "win_util.h"
+
+#define CHECK_VALID_HANDLE_WITH_RETURN_VALUE(win_handle, ret)         \
+    do {                                                              \
+        if ((win_handle) == NULL                                      \
+            || ((win_handle)->type == windows_handle_type_socket      \
+                && (win_handle)->raw.socket == INVALID_SOCKET)        \
+            || ((win_handle)->type == windows_handle_type_file        \
+                && (win_handle)->raw.handle == INVALID_HANDLE_VALUE)) \
+            return (ret);                                             \
+                                                                      \
+    } while (0)
+
+#define CHECK_VALID_HANDLE(win_handle) \
+    CHECK_VALID_HANDLE_WITH_RETURN_VALUE(win_handle, __WASI_EBADF)
+
+#define CHECK_VALID_FILE_HANDLE(win_handle)                        \
+    do {                                                           \
+        if ((win_handle) == NULL)                                  \
+            return __WASI_EBADF;                                   \
+                                                                   \
+        if ((win_handle)->type == windows_handle_type_socket)      \
+            return __WASI_EINVAL;                                  \
+                                                                   \
+        if (((win_handle)->type == windows_handle_type_file        \
+             && (win_handle)->raw.handle == INVALID_HANDLE_VALUE)) \
+            return __WASI_EBADF;                                   \
+                                                                   \
+    } while (0)
+
+#define CHECK_VALID_WIN_DIR_STREAM(win_dir_stream)         \
+    do {                                                   \
+        if ((win_dir_stream) == NULL)                      \
+            return __WASI_EINVAL;                          \
+        CHECK_VALID_FILE_HANDLE((win_dir_stream)->handle); \
+    } while (0)
+
+#define CHECK_VALID_PATH(path)          \
+    do {                                \
+        if (strstr(path, "\\") != NULL) \
```

Comment:
A little confused, windows path cannot contain '\\'?

Reply:
I'll add a comment to make it clearer - the issue is that the wasm application won't know which platform it's running on - on POSIX, backslashes are a valid filename character whereas on Windows, they are a path separator.

e.g. if the wasm app had the following code
```
FILE *file = fopen("foo\\bar", "r");
```
(assuming backslashes are allowed on Windows) this would open the`bar` file under `foo` rather than opening `foo\\bar` on POSIX which may or may not be what the user wanted.

So I thought the easiest option we have on Windows is to completely disallow backslashes in the path since it's probably not what the user intended (and there isn't a way in Windows to allow backslashes to be part of a filename). It's probably very much an edge case since it would be quite unusual to have a backslash in a filename anyway.

Reply:
hmm, I thought it's quite common to have a nested file structure and when the user wants to access a file in one of the (sub)directories, they need to provide a path to it (that includes `\`).

Reply:
Discussed with @loganek offline and we agreed to remove this check. My main concern was the differing behavior on POSIX and Windows re backslashes but I guess that's better than not allowing backslashes on windows at all.